### PR TITLE
feat: add Patreon connector

### PR DESCRIPTION
**Pull request by devtekve**
## Summary

This MR introduces the initial version of the Patreon connector for Logto, allowing users to authenticate via Patreon accounts. The implementation is based on the existing Hugging Face connector, adapting it to use Patreon's OAuth 2.0 authentication flow.

### Key Changes:
- Added a new `@logto/connector-patreon` package.
- Implemented OAuth 2.0 authentication for Patreon.
- Configured endpoints for fetching user details and handling tokens specific to Patreon.
- Updated README to reflect setup and configuration instructions for the Patreon connector.
- Acknowledged the base work from Silverhand Inc.'s Hugging Face connector and thanked @xiaoyijun for pointing it out.

## Testing

- Created unit tests to verify the OAuth token exchange process.
- Added integration tests to ensure the connector correctly interacts with Patreon's API.
- Manually tested the connector by authenticating with a real Patreon account, verifying that user details are retrieved correctly.
- Verified that errors are handled gracefully, including invalid credentials and denied access.

## Checklist

- [x] `.changeset` - Added a changeset file to record the new feature introduction.
- [x] unit tests - Developed unit tests for OAuth token exchange and user detail retrieval.
- [x] necessary TSDoc comments - Added TSDoc comments to the new methods and classes introduced for documentation.
- [ ] integration tests - Ran integration tests to ensure proper interaction with Patreon API.

### Note:
I am not familiar with the specifics on the integration tests for typescript... So for the integration test I ran `pnpm -r --parallel --workspace-concurrency=0 test:ci` but I have a hunch that's not enough or what we want because I did not see my connector on the list. 

That being said, I did test this locally manually. I am really hoping the ci/cd performs the necessary checks to ensure everything is good, otherwise I rely on other contributors / logto for just running another set of validations just in case... 

For maintainers: This code is yours, feel free to update and modify as you see fit.

### CODE REVIEW

File: packages/connectors/connector-patreon/src/types.ts [hunk 1]
```diff
@@ -0,0 +1,47 @@
...
+  data: z.object({
+    id: z.string(),
+    attributes: z.object({
+      full_name: z.string().optional().nullable(),
+      vanity: z.string().optional().nullable(),
+      url: z.string().optional().nullable(),
+      image_url: z.string().optional().nullable(),
+      email: z.string().optional().nullable(),
+      is_email_verified: z.boolean().optional(),
+      // Created: z.string().optional().nullable(), // Date as string, need to parse into date object if necessary
```
Comment (devtekve) [inline packages/connectors/connector-patreon/src/types.ts:34]:
```suggestion
```

Comment (devtekve) [inline packages/connectors/connector-patreon/src/types.ts:34]:
```suggestion
      is_email_verified: z.boolean().optional(),
```

**Comment by devtekve** [comment]
FYI I am force-pushing this to sign the commits as I saw it was asked for another similar PR https://github.com/logto-io/logto/pull/5886#issuecomment-2116862450

**Comment by xiaoyijun** [comment]
Hi @devtekve , great job! Thank you for your contribution.

I reviewed the code and noticed that Patreon also follows the OAuth2 protocol. This means we can implement it based on the `connector-oauth2` package, which will reduce the amount of code you need to write, you can refer to the implementation of the HuggingFace Connector (see https://github.com/logto-io/logto/pull/5797).

**Comment by devtekve** [comment]
hi @xiaoyijun , thanks for the review! Will take a look and ajust accordingly

**Comment by devtekve** [comment]
I've updated it now to follow Hugging Face's example @xiaoyijun thanks for the feedback.

File: packages/connectors/connector-patreon/src/constant.ts [hunk 2]
```diff
@@ -0,0 +1,52 @@
...
+    {
+      key: 'scope',
+      type: ConnectorConfigFormItemType.Text,
+      label: 'Scope',
+      required: false,
+      placeholder: '<scope>',
+      description:
+        "The `scope` determines permissions granted by the user's authorization. If you are not sure what to enter, do not worry, just leave it blank.",
+    },
+  ],
```
Comment (xiaoyijun) [inline packages/connectors/connector-patreon/src/constant.ts:49]: We can use predefined constants imported from `@logto/connector-oauth`.

```suggestion
import { clientIdFormItem, clientSecretFormItem, scopeFormItem } from '@logto/connector-oauth';

//...
formItems: [
    clientIdFormItem,
    clientSecretFormItem,
    {
      ...scopeFormItem,
      description:
        "The `scope` determines permissions granted by the user's authorization. If you are not sure what to enter, do not worry, just leave it blank.",
    },
  ],
```

File: .changeset/selfish-kangaroos-perform.md [hunk 3]
```diff
@@ -0,0 +1,5 @@
+---
+"@logto/connector-patreon": major
+---
+
+Adding a new patreon social connector for Logto
```
Comment (xiaoyijun) [inline .changeset/selfish-kangaroos-perform.md:5]:
```suggestion
add Patreon social connector
```

File: packages/connectors/connector-patreon/CHANGELOG.md [hunk 4]
```diff
@@ -0,0 +1,22 @@
...
+
+### Major Changes
+
+- Initial release of the Patreon connector.
+
+  This release introduces the Logto connector for Patreon, enabling social sign-in using Patreon accounts. It supports OAuth 2.0 authentication flow, fetching user information, and handling errors gracefully.
+
+- Base code adapted from the Hugging Face connector.
+
+  The Patreon connector's initial version is largely based on the Hugging Face connector developed by Silverhand Inc. and the Logto team. Many implementation patterns, error handling strategies, and configuration options were inspired by the Hugging Face connector, leveraging OAuth 2.0 which saved time and complexity. Acknowledgments to @xiaoyijun for pointing it out and to the original creators at Silverhand Inc. for their foundational work.
```
Comment (xiaoyijun) [inline packages/connectors/connector-patreon/CHANGELOG.md:13]: We can remove this part since it's better to focus on this connector.

```suggestion

```

**Review by xiaoyijun — CHANGES_REQUESTED** [review]
Great work @devtekve , I will have a test locally once you've update related codes.

**Comment by devtekve** [comment]
image
image

File: packages/connectors/connector-patreon/src/constant.ts (hunk 2 above)
Comment (devtekve) [reply packages/connectors/connector-patreon/src/constant.ts]: Thank you! Done!

File: .changeset/selfish-kangaroos-perform.md (hunk 3 above)
Comment (devtekve) [reply .changeset/selfish-kangaroos-perform.md]: DOne

**Comment by xiaoyijun** [comment]
Good job @devtekve , I just tested it locally, and everything works perfectly!

https://github.com/user-attachments/assets/0f518df3-ab4a-4e1b-a06a-29cbfa3ad237

**Review by xiaoyijun — APPROVED** [review]

File: packages/connectors/connector-patreon/CHANGELOG.md [hunk 5]
```diff
@@ -0,0 +1,18 @@
...
+- Initial release of the Patreon connector.
+
+  This release introduces the Logto connector for Patreon, enabling social sign-in using Patreon accounts. It supports OAuth 2.0 authentication flow, fetching user information, and handling errors gracefully.
+
+### Features
+
+- **OAuth 2.0 Authentication**: Support for OAuth 2.0 authentication flow with Patreon.
+- **User Information Retrieval**: Fetches user details such as full name, email, profile URL, and avatar.
+- **Error Handling**: Graceful handling of OAuth errors, including token exchange failures and user-denied permissions.
+- **Configurable Scope**: Allows customization of OAuth scopes to access different levels of user information.
```
Comment (gao-sun) [inline packages/connectors/connector-patreon/CHANGELOG.md:16]: this can be move to changeset and let the cli generate the file

Comment (xiaoyijun) [reply packages/connectors/connector-patreon/CHANGELOG.md]: updated.

**Review by gao-sun — APPROVED** [review]

**Pull request merged**